### PR TITLE
 Add support for tables `ContinuousBackup`

### PR DESCRIPTION
**Pull request by msvechla**
Issue #, if available: #1790

Description of changes:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by msvechla** [comment]
I'm able to set continuous backups to false and the backups are actually disabled on AWS side, however the controller keeps logging the following afterwards: 

> 2023-07-19T16:25:23.077Z        INFO    ackrt   desired resource state has changed      {"account": "404164987772", "role": "", "region": "eu-central-1", "kind": "Table", "namespace": "default", "name": "tenant-api-dynamodb-tenant-config", "is_adopted": false, "generation": 13, "diff": [{"Path":{"Parts":["Spec","ContinuousBackups","PointInTimeRecoveryEnabled"]},"A":false,"B":true}]}
2023-07-19T16:25:28.153Z        INFO    ackrt   desired resource state has changed      {"account": "404164987772", "role": "", "region": "eu-central-1", "kind": "Table", "namespace": "default", "name": "tenant-api-dynamodb-tenant-config", "is_adopted": false, "generation": 13, "diff": [{"Path":{"Parts":["Spec","ContinuousBackups","PointInTimeRecoveryEnabled"]},"A":false,"B":true}]}

Setting backups to enabled is not working at all at the moment

**Comment by a-hilaly** [comment]
>Setting backups to enabled is not working at all at the moment

This could be some eventual consistency of the DDB API. Maybe try the `Put` call that a quick `Get` call right after (using the CLI) to confirm?

**Comment by msvechla** [comment]
I found the issue, I was reading the wrong results from the `DescribeContinuousBackupOutput`:

```json
{
    "ContinuousBackupsDescription": {
        "ContinuousBackupsStatus": "ENABLED",
        "PointInTimeRecoveryDescription": {
            "PointInTimeRecoveryStatus": "DISABLED"
        }
    }
}
```

I was checking against `ContinuousBackupsStatus`, when I should have checked against `PointInTimeRecoveryStatus`.
Everything is working as expected now.

I also bumped the `acktest` dependency as suggested in the issue

**Comment by a-hilaly** [comment]
> I was checking against ContinuousBackupsStatus, when I should have checked against PointInTimeRecoveryStatus.
Everything is working as expected now.

@msvechla makes sense! Maybe we should name this field `PointInTimeRecoveryEnabled` and use the path `PointInTimeRecoverySpecification.PointInTimeRecoveryEnabled` instead?

**Comment by a-hilaly** [comment]
From the logs looks like the controller is struggling to understand the diff between latest and desired
```
2023-07-19T21:35:25.294Z	INFO	ackrt	desired resource state has changed	{"account": "021785965138", "role": "", "region": "us-west-2", "kind": "Table", "namespace": "default", "name": "table-lsi-7r1qmayaaekz3ycox3tiqw", "is_adopted": false, "generation": 1, "diff": [{"Path":{"Parts":["Spec","ContinuousBackups"]},"A":null,"B":{"pointInTimeRecoveryEnabled":false}}]}
```

**Comment by msvechla** [comment]
> > I was checking against ContinuousBackupsStatus, when I should have checked against PointInTimeRecoveryStatus.
> > Everything is working as expected now.
> 
> @msvechla makes sense! ~Maybe we should name this field `PointInTimeRecoveryEnabled` and use the path `PointInTimeRecoverySpecification.PointInTimeRecoveryEnabled` instead?~ `ContinuousBackup` is right here

Are you sure? I actually like just having `PointInTimeRecoveryEnabled`. What made you rethink this?

**Review by a-hilaly — COMMENTED** [review]
Thank you so much for working on this @msvechla - it's a great start! I left a few nit/comments below. 

Also, it would be amazing if we can also have an e2e test case for this feature (very similar to the TTL ones)

File: pkg/resource/table/hooks_continuous_backup.go
Comment (msvechla) [reply pkg/resource/table/hooks_continuous_backup.go:82]: This was the case for the first version I pushed. I since then fixed the issue and force-pushed to my branch. For me its no longer infinitely updating. I just get the log once and then the pitr is updated on the table

**Comment by a-hilaly** [comment]
>Are you sure? I actually like just having pointInTimeRecoveryEnabled as a field on the Kubernetes resource. What made you rethink this?
I have the refactoring ready, just let me know whether I should push

@msvechla I initially thought that the call was `PutPointInTime...` but I saw it was named `UpdateContinuousBackups`. I'm just thinking what if one day the dynamodb team adds a new field to `UpdateContinuousBackupsInput` struct do we want to keep adding new fields to the spec? or we keep everything encapsulated inside `ContinuousBackups` field. If you think that is very unlikely to happen i'm happy to go with a simple `PointInTimeRecoveryEnabled` field. 

Related to this topic: We are also working on [multi-version support](https://github.com/aws-controllers-k8s/community/pull/1780) (renaming, moving fields will be possible, with mutating webhooks)

File: pkg/resource/table/hooks.go
Comment (msvechla) [reply pkg/resource/table/hooks.go:453]: I think its not completely generated, there is the custom function defined in `generator.yaml`

```yaml
    update_operation:
      custom_method_name: customUpdateTable
```

I added the `if delta.DifferentAt ...` there, so this probably triggered the formatting of my editor

File: apis/v1alpha1/ack-generate-metadata.yaml
Comment (msvechla) [reply apis/v1alpha1/ack-generate-metadata.yaml:5]: sure, will do

File: pkg/resource/table/hooks.go
Comment (a-hilaly) [reply pkg/resource/table/hooks.go:453]: True true, i glitched! :)

**Comment by msvechla** [comment]
> > Are you sure? I actually like just having pointInTimeRecoveryEnabled as a field on the Kubernetes resource. What made you rethink this?
> > I have the refactoring ready, just let me know whether I should push
> 
> @msvechla I initially thought that the call was `PutPointInTime...` but I saw it was named `UpdateContinuousBackups`. I'm just thinking what if one day the dynamodb team adds a new field to `UpdateContinuousBackupsInput` struct, do we want to keep adding new fields to the spec? or we keep everything encapsulated inside `ContinuousBackups` field. If you think that is very unlikely to happen i'm happy to go with a simple `PointInTimeRecoveryEnabled` field.
> 
> Related to this topic: We are also working on [multi-version support](https://github.com/aws-controllers-k8s/community/pull/1780) (renaming, moving fields will be possible, with mutating webhooks)

Okay thats a valid point, then I will leave the spec as it is currently 👍

File: pkg/resource/table/hooks.go
Comment (a-hilaly) [inline pkg/resource/table/hooks.go:406]: The error we're seeing in the logs "can" be solved doing something like: https://github.com/aws-controllers-k8s/dynamodb-controller/blob/main/pkg/resource/table/hooks.go#L494-L498 - this tricks the delta function in believing that "nil" struct is equivalent to `spec.x = &{y: &false}`

File: apis/v1alpha1/ack-generate-metadata.yaml
Comment (a-hilaly) [reply apis/v1alpha1/ack-generate-metadata.yaml:5]: Sorry i meant `v0.26.1` (leatest version used the generate the controller)

File: pkg/resource/table/hooks_continuous_backup.go
Comment (a-hilaly) [reply pkg/resource/table/hooks_continuous_backup.go:82]: Was the table synced? (status.Conditions[ACKSynced = true])

Comment (msvechla) [reply pkg/resource/table/hooks_continuous_backup.go:82]: For me yes. This is the only log message I get when for example setting pitr from `enabled` to `disabled`:

```
2023-07-19T22:28:56.163Z        INFO    ackrt   desired resource state has changed      {"account": "223727272", "role": "", "region": "eu-central-1", "kind": "Table", "namespace": "default", "name": "my-dynamodb", "is_adopted": false, "generation": 27, "diff": [{"Path":{"Parts":["Spec","ContinuousBackups","PointInTimeRecoveryEnabled"]},"A":false,"B":true}]}
```

It is only displayed once. Table status shows:

```
Status:
  Ack Resource Metadata:
    Arn:               arn:aws:dynamodb:eu-central-1:404164987772:table/tenant-config
    Owner Account ID:  404164987772
    Region:            eu-central-1
  Conditions:
    Last Transition Time:  2023-07-19T22:29:01Z
    Message:               Resource synced successfully
    Reason:
    Status:                True
    Type:                  ACK.ResourceSynced
```

File: apis/v1alpha1/ack-generate-metadata.yaml
Comment (msvechla) [reply apis/v1alpha1/ack-generate-metadata.yaml]: Ah I just updated it to `v0.26.0`. Will try to continue tomorrow and update to the desired one

**Comment by msvechla** [comment]
> Thank you so much for working on this @msvechla - it's a great start! I left a few nit/comments below.
> 
> Also, it would be amazing if we can also have an e2e test case for this feature (very similar to the TTL ones)

I have added some e2e tests, maybe you can take a look. Otherwise my manual tests also worked great now

File: pkg/resource/table/hooks.go
Comment (a-hilaly) [inline pkg/resource/table/hooks.go:563]: :+1:

File: pkg/resource/table/hooks_continuous_backup.go
Comment (a-hilaly) [inline pkg/resource/table/hooks_continuous_backup.go:41]: Not sure how this API works, but if a user deletes the field `ContinuousBackups` should we also explicitly set `SetPointInTimeRecoveryEnabled` to false? thinking about something like below, thoughts?
```go
    pitrSpec.SetPointInTimeRecoveryEnabled(*false)
    // only set the PITR is a user explicitly passes the full struct with enabled=true
	if desired.ko.Spec.ContinuousBackups != nil &&
		desired.ko.Spec.ContinuousBackups.PointInTimeRecoveryEnabled != nil {
		pitrSpec.SetPointInTimeRecoveryEnabled(
			*desired.ko.Spec.ContinuousBackups.PointInTimeRecoveryEnabled,
		)
	}
```

**Review by a-hilaly — COMMENTED** [review]
This is great @msvechla! I have two nits below and happy to merge right away

File: pkg/resource/table/hooks_continuous_backup.go
Comment (msvechla) [reply pkg/resource/table/hooks_continuous_backup.go:41]: the `pitrSpec.SetPointInTimeRecoveryEnabled(*false)` is already happening implicitly. If pitr is enabled and the user removes the entire `ContinuousBackups` field, then pitr is already disabled today

**Review by a-hilaly — APPROVED** [review]
Dankeshoen @msvechla ! :)
/lgtm

File: pkg/resource/table/hooks.go
Comment (msvechla) [reply pkg/resource/table/hooks.go]: fixed

**Comment by msvechla** [comment]
> Can you please send a release PR once this is merged? we can go for `1.2.0`

Thanks for your guidance, will do. Can I just follow this example by bumping the versions in those files manually? https://github.com/aws-controllers-k8s/dynamodb-controller/pull/72/files

Or is there some tooling around bumping the version?

**Comment by a-hilaly** [comment]
>Thanks for your guidance, will do. Can I just follow this example by bumping the versions in those files manually? https://github.com/aws-controllers-k8s/dynamodb-controller/pull/72/files

Exactly! Those files can be regenerated with new versions using something like `RELEASE_VERSION=v1.0.3 ./scripts/build-controller-release.sh dynamodb` (from the code-generator repository) - full documentation is here https://aws-controllers-k8s.github.io/community/docs/contributor-docs/release/

Once the release PR is merged prow will take care of publishing new images/helm charts

**Comment by msvechla** [comment]
Thanks, created a release PR here: https://github.com/aws-controllers-k8s/dynamodb-controller/pull/78

**Pull request merged**
